src/Engine/ProtoCore/FFI/CLRObjectMarshler.cs

```diff
@@ -304,6 +316,38 @@ private static Type GetElementType(Type expectedCLRType)
             return elementType;
         }
 
+        #region DS_ARRAY_TO_CS_DICTIONARY
+
+        protected object ToDictionary(StackValue dsObject, ProtoCore.Runtime.Context context, Interpreter dsi, System.Type keyType, System.Type valueType)
+        {
+            if (!dsObject.IsArray)
+            {
+                return null;
+            }
+
+            var dict = (IDictionary)Activator.CreateInstance(typeof(Dictionary<,>).MakeGenericType(keyType, valueType)); 
+            var dsKeys = ArrayUtils.GetKeys(dsObject, dsi.runtime.RuntimeCore);
+
+            foreach (var dsKey in dsKeys)
+            {
+                var key = primitiveMarshaler.UnMarshal(dsKey, context, dsi, keyType);
+                if (key == null || !keyType.IsAssignableFrom(key.GetType()))
```

Comment:
The marshaller's behavior is, if the DS data is an array, then skips the expected type and will use collection marshaller. Is it expected?